File: jugglerhelper/src/main/java/net/maxsmr/jugglerhelper/fragments/FragmentFinder.java
package net.maxsmr.jugglerhelper.fragments;

import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.util.Pair;

import net.maxsmr.commonutils.android.gui.adapters.CustomFragmentStatePagerAdapter;
import net.maxsmr.commonutils.data.CompareUtils;
import net.maxsmr.commonutils.data.Predicate;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;

public final class FragmentFinder {

    private FragmentFinder() {
        throw new AssertionError("No instances");
    }

    @Nullable
    public static Pair<Integer, Fragment> findFragmentById(@Nullable FragmentManager fm, int id) {
        if (fm != null) {
            return findFragmentById(fm.getFragments(), id);
        }
        return null;
    }

    @Nullable
    public static Pair<Integer, Fragment> findFragmentById(@Nullable Collection<Fragment> fragments, int id) {
        return Predicate.Methods.findWithIndex(fragments, fragment -> fragment != null && !fragment.isDetached() && fragment.getId() == id);
    }

    @Nullable
    public static Pair<Integer, Fragment> findFragmentByTag(FragmentManager fm, String tag) {
        if (fm != null) {
            return findFragmentByTag(fm.getFragments(), tag);
        }
        return null;
    }

    @Nullable
    public static Pair<Integer, Fragment> findFragmentByTag(@Nullable Collection<Fragment> fragments, String tag) {
        return Predicate.Methods.findWithIndex(fragments, fragment ->   fragment != null && !fragment.isDetached() && CompareUtils.stringsEqual(fragment.getTag(), tag, false));

    }

    @Nullable
    @SuppressWarnings("unchecked")
    public static <F extends Fragment> Pair<Integer, F> findFragmentByClass(@Nullable FragmentManager fm, @Nullable Class<F> fragmentClass) {
        if (fm != null) {
            return findFragmentByClass(fm.getFragments(), fragmentClass);
        }
        return null;
    }

    @Nullable
    @SuppressWarnings("unchecked")
    public static <F extends Fragment> Pair<Integer, F> findFragmentByClass(@Nullable Collection<Fragment> fragments, @Nullable Class<F> fragmentClass) {
        Pair<Integer, Fragment> result = null;
        if (fragmentClass != null) {
            result = Predicate.Methods.findWithIndex(fragments, fragment -> fragment != null && !fragment.isDetached() && fragmentClass.isAssignableFrom(fragment.getClass()));
        }
        if (result != null) {
            return new Pair<>(result.first, (F) result.second);
        }
        return null;
    }

    @Nullable
    @SuppressWarnings("unchecked")
    public static <F extends Fragment> Pair<Integer, F> findFragmentByInstance(@Nullable FragmentManager fm, @Nullable F targetFragment) {
        if (fm != null) {
            return findFragmentByInstance(fm.getFragments(), targetFragment);
        }
        return null;
    }

    @Nullable
    @SuppressWarnings("unchecked")
    public static <F extends Fragment> Pair<Integer, F> findFragmentByInstance(@Nullable Collection<Fragment> fragments, @Nullable F targetFragment) {
        Pair<Integer, Fragment> result = Predicate.Methods.findWithIndex(fragments, fragment -> fragment != null && !fragment.isDetached() &&
                CompareUtils.objectsEqual(fragment, targetFragment) &&
                (targetFragment == null || targetFragment.getClass().isAssignableFrom(fragment.getClass())));
        if (result != null) {
            return new Pair<>(result.first, (F) result.second);
        }
        return null;
    }

    // TODO update CustomFragmentStatePagerAdapter
    @Deprecated
    @Nullable
    public static <F extends Fragment> Pair<Integer, F> findFragmentByClass(@NotNull Class<F> fragmentClass, @Nullable CustomFragmentStatePagerAdapter pagerAdapter) {
        Pair<Integer, F> result = null;
        if (pagerAdapter != null) {
            for (int index = 0; index < pagerAdapter.getCount(); index++) {
                Fragment fragment = pagerAdapter.getFragmentInstance(index);
                if (fragmentClass.isAssignableFrom(fragment.getClass())) {
                    //noinspection unchecked
                    result = new Pair<>(index, (F) fragment);
                    break;
                }
                index++;
            }
        }
        return result;
    }
}

File: jugglerhelper/src/main/kotlin/net/maxsmr/jugglerhelper/utils/FragmentFinder.kt
package net.maxsmr.jugglerhelper.utils

import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager

fun <F : Fragment> findFragment(fragmentManager: FragmentManager, params: FragmentSearchParams<F>): Pair<Int, Fragment>? =
        findFragment(fragmentManager.fragments, params)

fun <F : Fragment> findFragment(fragments: Collection<Fragment?>, params: FragmentSearchParams<F>): Pair<Int, Fragment>? =
        filterFragments(fragments, params).toList().getOrNull(0)

fun <F : Fragment> filterFragments(fragmentManager: FragmentManager, params: FragmentSearchParams<F>): Map<Int, Fragment> =
        filterFragments(fragmentManager.fragments, params)

@Suppress("UNCHECKED_CAST")
fun <F : Fragment> filterFragments(fragments: Collection<Fragment?>, params: FragmentSearchParams<F>): Map<Int, Fragment> {
    val result = mutableMapOf<Int, Fragment>()
    val filteredFragments: List<Pair<Int, Fragment>> = fragments
            .mapIndexed { index, fragment -> Pair(index, fragment) }
            .filter { fragmentPair ->
                val fragment = fragmentPair.second
                with(params) {
                    var isMatch = false
                    if (fragment != null && !fragment.isDetached) {
                        isMatch = true
                        if (id != null && id != fragment.id) {
                            isMatch = false
                        }
                        if (tag != null && tag != fragment.tag) {
                            isMatch = false
                        }
                        if (fragmentClass != null && !fragmentClass.isAssignableFrom(fragment.javaClass)) {
                            isMatch = false
                        }
                        if (this.fragment != null && this.fragment != fragment) {
                            isMatch = false
                        }
                        if (additionalPredicate != null && !additionalPredicate.invoke(fragment)) {
                            isMatch = false
                        }
                    }
                    isMatch
                }
            }.mapNotNull {
                if (it.second == null) {
                    return@mapNotNull null
                }
                it
            } as List<Pair<Int, Fragment>>
    filteredFragments.forEach {
        result[it.first] = it.second
    }
    return result
}

data class FragmentSearchParams<F : Fragment>(
        val id: Int? = null,
        val tag: String? = null,
        val fragmentClass: Class<F>? = null,
        val fragment: F? = null,
        val additionalPredicate: ((Fragment) -> Boolean)? = null
)